src/ir/module-splitting.cpp

```diff
@@ -90,81 +88,118 @@ namespace {
 
 template<class F> void forEachElement(Table& table, F f) {
   for (auto& segment : table.segments) {
-    assert(segment.offset->is<Const>() &&
-           "TODO: handle non-constant segment offsets");
-    Index baseOffset = segment.offset->cast<Const>()->value.geti32();
+    Name base = "";
+    Index offset = 0;
+    if (auto* c = segment.offset->dynCast<Const>()) {
+      offset = c->value.geti32();
+    } else if (auto* g = segment.offset->dynCast<GlobalGet>()) {
+      base = g->name;
+    }
     for (size_t i = 0; i < segment.data.size(); ++i) {
-      f(Index(baseOffset + i), segment.data[i]);
+      f(base, offset + i, segment.data[i]);
     }
   }
 }
 
 struct TableSlotManager {
+  struct Slot {
+    // If `global` is empty, then this slot is at a statically known index.
+    Name global;
+    Index index;
+
+    // Generate code to compute the index of this table slot
+    Expression* makeExpr(Module& module);
+  };
   Module& module;
   Table& table;
   Table::Segment* activeSegment = nullptr;
-  Index activeBase = 0;
-  std::map<Name, Index> funcIndices;
+  Slot activeBase = {"", 0};
+  std::map<Name, Slot> funcIndices;
 
   TableSlotManager(Module& module);
 
   // Returns the table index for `func`, allocating a new index if necessary.
-  Index getIndex(Name func);
-  void addIndex(Name func, Index index);
+  Slot getSlot(Name func);
+  void addSlot(Name func, Slot slot);
 };
 
-void TableSlotManager::addIndex(Name func, Index index) {
-  auto it = funcIndices.insert(std::make_pair(func, index));
-  assert(it.second && "Function already has multiple indices");
+Expression* TableSlotManager::Slot::makeExpr(Module& module) {
+  Builder builder(module);
+  if (global) {
+    Expression* getBase = builder.makeGlobalGet(global, Type::i32);
+    return index == 0 ? getBase
+                      : builder.makeBinary(
+                          AddInt32, getBase, builder.makeConst(int32_t(index)));
+  } else {
+    return builder.makeConst(int32_t(index));
```

Comment:
can deduplicate the `makeConst` into a local?

Reply:
Not without leaking it in cases where it is not used, but perhaps that doesn't matter so much. For now I'll deduplicate it into a small lambda to avoid leaks. Let me know what you think; happy to iterate further.

Reply:
Ah, I missed the leak issue...

I guess it's the same amount of code. Reading the lambda, I don't have a preference, so lgtm either way.